### PR TITLE
Firefox: Open In Browser, uBlock, Disable bookmarks

### LINKED ISSUE #105: better link handling from navigation-off webapps
When I use a web app with navigation bar off, I can't click on any links without screwing everything up. Take Google Calendar or Google Messages, for example: they make sense as a navbar-off web app. But what happens when I need to click on a Zoom link? It opens in the same window. There's no visible back button. Sometimes I'm able to keyboard-shortcut back, sometimes not (not sure why); I often just close and re-open the app to get back to the original page (e.g. calendar.google.com).

I get that I can obviously just turn the navigation bar on. But I definitely prefer the streamlined, clutter-free look with the navbar off.

I'm not sure what the ideal way to fix this is. Here are some ideas:

- Switch to navbar-on when I click a link, and open as tab in same window. Remove navbar again if I close all but the original tab.
- Open all links as popups, each in their own window.
- Open links in an existing copy of the same browser under the default profile, if running. This is suggested in [a related thread](https://github.com/linuxmint/webapp-manager/issues/14#issuecomment-688999373).
- Expose a back button so I can at least get back to where I started. (Or perhaps a "reload" button that restores the original page.)
- Prompt user to choose where to open the link. (Probably way too annoying.)
- Allow the user to configure the behavior, per web app.

This is the one issue that seriously impairs the usability of the web apps for me. I made a bunch of them and quit using most of them because dealing with links is too annoying.

Note that this issue is raised in [a related thread](https://github.com/linuxmint/webapp-manager/issues/14#issuecomment-688999373), but I think that it deserves its own discussion here.

### PR DESCRIPTION
**Pull request by Saghen**
# Description

This PR allows Firefox to open URLs on your default browser, adds uBlock as a default extension and disables the bookmark bar by default. Opening URLs remotely was achieved via

- Adding an extension which intercepts clicks on `<a>` tags, `window.open` and any new tab and redirects them to `browser:${url}` if they're not on the current domain
- Adding an `extensions.json` file to pre-load extensions without the user having to manually enable them
- Adding a custom protocol `browser:` by adding `network.protocol-handler.expose.browser: false` in `user.js`
- Defining `handlers.json` pointing to `browser:` to call `/usr/share/webapp-manager/firefox/open-browser-without-env.sh`
- `open-browser-without-env.sh` removes ENV variables that cause Firefox to fail to open and calls `xdg-open` to open the URL in the default browser

Couple notes for future work:

There's very likely flaws in the approach for the extension. The code checks if the `event.target` is an `<a>` tag but does not check if it's an element nested inside of an `<a>` tag. Testers needed!

The addition of `extensions.json` may be a fragile approach to adding an extension by default but it's the only working approach I found. Testers needed!

Closes #105

### CODE REVIEW

**Review by hsbasu — CHANGES_REQUESTED** [review]
Please do not move the scripts or rename them. They are there for a reason.

**Comment by hsbasu** [comment]
Please split the commits. Each commit should contain only a single major change. If you are including multiple minor changes in the commits create the commit message like the following

```
summary of major change and some minor changes

- describe the major change
- filename1: minor change 1, minor change 2
- filename 2: minor change 1, minor change 2
```

**Comment by hsbasu** [comment]
I would highly discourage suppying binary files like .xpi along with an open source package. There should be a more elegant way to achieve this.

**Comment by Saghen** [comment]
Indeed I would too but I wasn't able to find a way to get Firefox to load the extension by default without resorting to signing it. I'll take another crack at it today to see if there's an alternative

**Comment by Saghen** [comment]
Unfortunately, after testing and doing some research, it looks like it's only possible to load unsigned extensions permanently in the development and nightly releases of Firefox.

From https://wiki.mozilla.org/Add-ons/Extension_Signing#FAQ
image

**Comment by hsbasu** [comment]
If this is the case, then it might be a security issue for the users. I have no knowledge on internet securities though.

**Comment by bkmo** [comment]
These commits work fine on my fork but I am not sure of the default browser implementation, Maybe it would be nicer to have a right-click menu item for the link that selects "open in default browser". That way it is not an all or nothing approach. It is also redirecting addons.mozilla.org to the default browser and breaking addon installs from that page.

**Comment by Saghen** [comment]
Definitely sounds like I'll need to whitelist the addons site like I did with the `about:` pages. The default behavior for Chromium is to open links in the default browser so this PR intended to match that behavior. What's your use case for not wanting that behavior? I imagine this is something we could make configurable

**Comment by bkmo** [comment]
I have an hidden/dropdown Nav Bar enabled via userChrome.css. If the page opens in the app then I have a tab bar that shows to
close it, and then the tab bar disappears if it was the last tab. (check out my fork of your repo : https://github.com/bkmo/webapp-manager-bkmo/tree/firefox-improvements   The drawback of opening in a new tab is that if the site sets cookies for login then I have to login to the site if not already done. In the default browser this would probably log me in. Replying to your comment from my email opened in a new tab with me logged out of Github. A right click-open in default browser is one option. As is being able to whitelist sites to open in a new tab. I will look into whitelisting addons.mozilla.org and give it another go.

**Comment by bkmo** [comment]
I tried with my limited knowledge to whitelist addons.mozilla.org, but failed. I was toying around with background.js. Do I need to make the extension with the files in the firefox-extension directory? I assume so as it is outside the app.

**Comment by bkmo** [comment]
My attempts at adding an addons.mozilla.org exception to the extension have miserably failed. I'll leave it up to you for assistance in creating the extension.

File: firefox-extension/background.js
Comment (SantoJambit) [inline firefox-extension/background.js:1]: This will stop working when manifest v3 is in place, as the script will be shut down and recreated when needed.
If you need to store something, use session storage: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/session

File: firefox-extension/background.js
Comment (SantoJambit) [inline firefox-extension/background.js:13]: You might want to take a look at this instead:
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webNavigation/onBeforeNavigate

**Review by SantoJambit — COMMENTED** [review]
Hi, I'm somewhat experienced on webextension development, so I thought I'd give some feedback on this approach.

I don't like the idea of extensions being installed by default, especially ones that are not the official ones and would have to be updated by the maintainer of this repo on a regular basis.

I wonder if it would be better to make this a published webextension instead and let the user install it via the addons manager. That works with Firefox (tried with ublock), as you can use alt to open the menu. For Chromium based browsers a different approach might be needed.

Maybe you could add another launch button, which starts the apps profile on a specific page (the addons store for the used browser).

About addons.mozilla.org: webextensions are not allowed to work on that url for security reasons.

File: usr/share/webapp-manager/firefox/profile/handlers.json
Comment (SantoJambit) [inline usr/share/webapp-manager/firefox/profile/handlers.json:18]: I personally don't use gmail, so if this suddenly started running gmail for me, I'd be upset.

File: firefox-extension/background.js
Comment (bkmo) [reply firefox-extension/background.js:1]: MV3 should not be a problem on Firefox correct?

Comment (SantoJambit) [reply firefox-extension/background.js:1]: Firefox will also implement MV3. There are exceptions, but not related to this part.
See: https://blog.mozilla.org/addons/2022/06/08/manifest-v3-firefox-developer-preview-how-to-get-involved/

**Comment by Saghen** [comment]
> I wonder if it would be better to make this a published webextension instead and let the user install it via the addons manager. 

@SantoJambit I don't mind this approach but because the `handlers.json` file must be placed in the Firefox profile, we'll also have to provide a profile with it for people who wish to use the extension and without it for those who don't.

To summarize, the following changes will be required:
- [ ] Rework the extension for Service Worker
- [ ] Revisit using `onBeforeNavigate` for intercepting navigation
- [ ] Ensure no excess handlers/configs are present in the profile
- [ ] Provide a configuration option for enabling the `handlers.json` and immediately open the browser to the addon's page for manual user installation when enabled
- [ ] Publish the extension on addons.mozilla.org presumably under a Linux Mint account

Nice to have:
- [ ] Provide a button for opening directly to the addons
- [ ] Provide a means to get to the extension configuration options when the navigation bar is hidden

I'm no longer using WebApp Manager so my motivation to work on this is low. Would be appreciated if anyone interested in collaborating on this emailed me at liam@super.fish

**Comment by SantoJambit** [comment]
I'm not using WebApp Manager, partly because of this missing feature, but I don't really have the time to work on (and support) yet another project.

I was just looking at the repo for a solution, found this issue and thought I'd give some advise.

Honestly, I only use webapps on my work computer and there I don't care as much about what browser I used. The built-in functionality of chromium to [create webapps](https://support.google.com/chrome_webstore/answer/3060053?hl=en) seems to work good enough (it even has a couple of more options). The only downside is that external links don't open in the default browser, which made me switch my default browser from Firefox to Chromium on my work computer. I wouldn't do that on my private computer, but there I don't use webapps anyway.

**Comment by Saghen** [comment]
If someone is interested in continuing this PR, they can fork from my branch and open a new PR

**Pull request closed without merging**
